Threadsafety login/logout/cancellogout and auto logout for lost addclasss bots

hermensbas opened the pull request:
Modified version of https://github.com/mod-playerbots/mod-playerbots/pull/1722
Subreference: https://github.com/mod-playerbots/mod-playerbots/issues/1669

Should also fix multiple logout crashes.

All the credits to @Wishmaster117 

- No direct logouts then via queue to prevent faulty status during strategy handling downstream
- AddClass bots removed when no group or group without the master

We dont care whether they are in LF queues or BG for that matter, thats the responsibility of the owner/master

Tested:
- Removing addClass by leave command and whether the bots logs out
- Multiple addClass in group and raid format
- Leave command during LFG group, dungeon or battleground
- Add class bots and logout and login with master and whether they remain in the group 



(will update after i am done and done testing)

Wishmaster117 commented:
> Modified version of #1722 Subreference: #1669 Should also fix multiple logout crashes.
> 
> All the credits to @Wishmaster117
> 
> * No direct logouts then via queue to prevent faulty status in strategy handling
> * AddClass bots removed when no group or group without the master
> 
> We dont care whether they are in LF queues or BG for that matter, thats the responsibility of the owner/master
> 
> Tested:
> 
> * Removing addClass by leave command, logs out.
> * Multiple addClass in group and raid format
> * Leave command during LFG group, dungeon or battleground
>   -Add class bots and logout and login and whether they remain in the group

Hi bro, 
Can you put a tag Multibot need update to remember me to add login functionality in Multibot uninvite button please 🙏?

hermensbas commented:
> Hi, as we are using logout queue maybe it's better tu remove the bots from the queue in cancelogout? ... PlayerbotAI* botAI = GET_PLAYERBOT_AI(bot); if (!botAI || botAI->IsRealPlayer()) continue;
> 
> // NEW: remove this bot from the logout queue iif present. { auto endIt = std::remove(m_pendingLogout.begin(), m_pendingLogout.end(), bot->GetGUID()); if (endIt != m_pendingLogout.end()) { m_pendingLogout.erase(endIt, m_pendingLogout.end()); LOG_DEBUG("playerbots", "CancelLogout: removed '{}' from pending logout queue", bot->GetName().c_str()); } }
> 
> ```
> 
> And
> 
> ...
> GET_PLAYERBOT_AI(bot);
>          if (!botAI || botAI->IsRealPlayer())
>              continue;
>  
>          if (botAI->GetMaster() != master)
>              continue;
>  
> // NEW: remove this  bot from the logout queue iif present.
>         {
>             auto endIt = std::remove(m_pendingLogout.begin(), m_pendingLogout.end(), bot->GetGUID());
>             if (endIt != m_pendingLogout.end())
>             {
>                 m_pendingLogout.erase(endIt, m_pendingLogout.end());
>                 LOG_DEBUG("playerbots", "CancelLogout: removed '{}' from pending logout queue", bot->GetName().c_str());
>             }
>         }
> ```

aah yes good call!

hermensbas commented:
I just realize these functions are called by multiple threads, world, network, map so i am gonna add logout, cancelogout and logoutAll and make them thread-safe. Login in threadsafe on itself, so doesnt require any changes.

hermensbas commented:
Actually based on Playerbots script hooks they are called Through Network, World and Map threads depending on context. So i am gonna wrap both cancel, logout and make the logout all threadSafe. Which is this mutex would be a good approach.

hermensbas commented:
As in:

- OnPacketReceived <- NetWork thread
- OnPlayerbotLogout, OnPlayerbotLogoutBots <- world thread
- HandlePlayerbotCommand in PlayerAI (which invokes logout as well) -> map thread
